887 standardise variants (#995)
Dyalog/documentation: language-reference-guide/docs/system-functions/xml.md

`⎕XML` converts an XML string into an APL array or converts an APL array into an XML string.

> Options for `⎕XML` are specified using the _variant_ operator `⍠` or by the optional left argument `X`. The former is recommended but the older mechanism using the left argument is still supported.

Reviewer: ```suggestion
Options for `⎕XML` are specified using the _variant_ operator `⍠` or by the optional left argument `X`. The former is recommended but the optional left argument is supported.
```

Attribute values can be delimited by either double quotes as shown or single quotes (apostrophes); they cannot contain certain characters (the delimiting quote, '&' or '<') and these must be represented by entity or character references.

The content of elements can be zero or more mixed occurrences of character data and nested elements. Tags and attribute names *describe* data, attribute values and the content within tags contain the data itself. Nesting of elements allows structure to be defined.

Because certain markup which describes the format of allowable data (such as element type declarations and attribute-list declarations) is not processed, no error will be reported if element contents and attributes do not conform to their restricted declarations, nor are attributes automatically added to tags if not explicitly given.

Attributes with names beginning **xml:** are reserved. Only **xml:space** is treated specially by `⎕XML`. When converting both from and to XML, the value for this attribute has the following effects on space normalization for the character data within this element and child elements within it (unless subsequently overridden):

> - **default** – space normalization is as determined by the `Whitespace` option. 

Reviewer: ```suggestion
- **default** – space normalisation is as determined by the `Whitespace` variant option. 
```

Then, character references and entity references are emitted in place of characters where necessary, to ensure that valid XML is generated. However, markup, if present, is *not* validated and it is possible to generate invalid XML if care in not taken with markup constructs.

## Variant Options

> `⎕XML` supports three variant options, `Whitespace`, `Markup` and `UnknownEntity`, summarised in [](#variantoptionsforxml) and described in detail beneath it. There is no principal option. Each option can also be specified through the optional left argument `X` (retained for backwards compatibility) using the alternative name shown in the table; the names are case-sensitive and must be spelled exactly as shown.

Reviewer: ```suggestion
`⎕XML` supports three variant options, `Whitespace`, `Markup`, and `UnknownEntity`, summarised in [](#variantoptionsforxml) and described in detail beneath it. There is no principal option. Each option can also be specified through the optional left argument `X` (retained for backwards compatibility) using the alternative name shown in [](#variantoptionsforxml); the names are case-sensitive and must be written exactly as shown.
```

Table: Variant options for `⎕XML` { #variantoptionsforxml }

|Variant Option|Left-argument name|Valid Values|Default|Effect|
|---|---|---|---|---|

> |[`Whitespace`](#variant-option-whitespace)|`whitespace`|`Strip`, `Trim` or `Preserve`|`Strip`|Handling of white space in character data.|

Reviewer: ```suggestion
|[`Whitespace`](#variant-option-whitespace)|`whitespace`|`Strip`, `Trim`, or `Preserve`|`Strip`|Handling of white space in character data.|
```

Errors detected in the input arrays or options will all cause `DOMAIN ERROR`.

### Variant Option: `Whitespace`

> The left-argument name is `whitespace`. When converting from XML, `Whitespace` specifies the default handling of white space surrounding and within character data. When converting to XML `Whitespace` specifies the default formatting of the XML. Note that attribute values are not comprised of character data so white space in attribute values is always preserved.

Reviewer: ```suggestion
The left-argument name is `whitespace`.

When converting from XML, `Whitespace` specifies the default handling of white space surrounding and within character data. When converting to XML `Whitespace` specifies the default formatting of the XML. Note that attribute values are not comprised of character data, so white space in attribute values is always preserved.
```

└∊──────────────────────────────────────┘
```

### Variant Option: `Markup`

> The left-argument name is `markup`. When converting from XML, `Markup` determines whether markup (other than entity tags) appears in the output array or not. When converting to XML `Markup` has no effect.

Reviewer: ```suggestion
The left-argument name is `markup`.

When converting from XML, `Markup` determines whether markup (other than entity tags) appears in the output array. When converting to XML `Markup` has no effect.
```

└∊──────────────────────────────────────────────┘
```

### Variant Option: `UnknownEntity`

> The left-argument name is `unknown-entity`. When converting from XML, this option determines what happens when an unknown entity reference, or a character reference for a Unicode character which cannot be represented as an APL character, is encountered. In Classic versions of Dyalog APL that is any Unicode character which does not appear in `⎕AVU`. When converting to XML, this option determines what happens to Esc characters (`⎕UCS 27`) in data.

Reviewer: ```suggestion
The left-argument name is `unknown-entity`.

When converting from XML, this option determines what happens when an unknown entity reference, or a character reference for a Unicode character that cannot be represented as an APL character, is encountered. In Classic versions of Dyalog APL this is any Unicode character that does not appear in `⎕AVU`. When converting to XML, this option determines what happens to Esc characters (`⎕UCS 27`) in data.
```
